src/Functions/URL/ExtractFirstSignificantSubdomain.h

```diff
@@ -119,40 +122,63 @@ struct ExtractFirstSignificantSubdomain
 
         const auto * begin = tmp;
         const auto * end = begin + domain_length;
-        const char * last_2_periods[2]{};
-        const char * prev = begin - 1;
+        const char * last_periods[2]{};
+        last_periods[0] = begin - 1;
+        StringRef excluded_host{};
 
         const auto * pos = find_first_symbols<'.'>(begin, end);
         while (pos < end)
         {
-            if (lookup(pos + 1, end - pos - 1))
+            size_t host_len = static_cast<size_t>(end - pos - 1);
+            StringRef host{pos + 1, host_len};
+            TLDType tld_type = lookup(host);
+            switch (tld_type)
             {
-                res_data += prev + 1 - begin;
-                res_size = end - 1 - prev;
-                return;
+                case TLDType::TLD_NONE:
+                    break;
+                case TLDType::TLD_REGULAR:
+                    res_data += last_periods[0] + 1 - begin;
+                    res_size = end - 1 - last_periods[0];
+                    return;
+                case TLDType::TLD_ANY:
+                {
+                    StringRef regular_host{last_periods[0] + 1, static_cast<size_t>(end - 1 - last_periods[0])};
+                    if (last_periods[1] && excluded_host != regular_host)
+                    {
+                        /// Return TLD_REGULAR + 1
+                        res_data += last_periods[1] + 1 - begin;
+                        res_size = end - 1 - last_periods[1];
+                    }
+                    else
+                    {
+                        /// Same as TLD_REGULAR
+                        res_data += last_periods[0] + 1 - begin;
+                        res_size = end - 1 - last_periods[0];
+                    }
+                    return;
+                }
+                case TLDType::TLD_EXCLUDE:
+                    excluded_host = host;
```

Comment:
> An exclamation mark (!) at the start of a rule marks an exception to a **previous** wildcard rule. An exception rule takes priority over any other matching rule.

I'm not sure what it meant, what if you have
```
*.kawasaki.jp
.
.
.
!city.kawasaki.jp
```
Should the exception apply here?
I can't think of any case where that would produce something valid but just to be sure.

Reply:
Yes. It simply means that city.kawasaki.jp is not tld, while everything else is, for example `foo.kawasaki.jp`.
Actually there is an example in the tests:

```sql
select cutToFirstSignificantSubdomainCustom('foo.kawasaki.jp', 'public_suffix_list');
--foo.kawasaki.jp
select cutToFirstSignificantSubdomainCustom('foo.foo.kawasaki.jp', 'public_suffix_list');
--foo.foo.kawasaki.jp
select cutToFirstSignificantSubdomainCustom('city.kawasaki.jp', 'public_suffix_list');
--city.kawasaki.jp
select cutToFirstSignificantSubdomainCustom('some.city.kawasaki.jp', 'public_suffix_list');
--city.kawasaki.jp
```

Reply:
Sorry, I notice that the bolded `previous` isn't obvious. I meant, should the exception be right after the wildcard rule or it can be anytime after it?